### PR TITLE
feat(orchestrator): add reference_output to quantize_block to skip collect_reference

### LINKED ISSUE #2288: quantize_block always re-runs collect_reference even when caller has FP16 outputs cached
# intel/auto-round Issue: Add `reference_output` parameter to `quantize_block` to allow callers to skip the collect_reference forward pass

## Summary

`CompressionOrchestrator.quantize_block` always runs an internal FP16 forward pass
("collect_reference / Step 3") to obtain reference outputs for the SignSGD loss
computation. This is redundant when the caller already holds the identical tensors
from a prior pass. Exposing a `reference_output=None` parameter would allow callers
to skip the forward pass entirely, saving significant CPU RAM and wall-clock time for
large models.

---

## Context: Integration with LLM-Compressor's SequentialPipeline

[vllm-project/llm-compressor](https://github.com/vllm-project/llm-compressor) uses auto-round
via `AutoRoundModifier`, which calls `quantize_block` once per transformer block. The
`SequentialPipeline` runs layer-by-layer and caches each subgraph's FP16 output in an
`IntermediatesCache` to feed as input to the next subgraph.

When `propagate_error=False` (the auto-round default), the same unquantized inputs flow
to every subgraph. Therefore:

```
IntermediatesCache[k+1]  =  FP16 output of block k (unquantized inputs, unquantized weights)
collect_reference output =  FP16 output of block k (unquantized inputs, unquantized weights)
```

Same block + same input + same weights → **outputs are mathematically identical**.
`collect_reference` re-derives them from scratch, duplicating the memory.

---

## Measured Impact

For Qwen3.5-MoE 122B (hidden=7168, seq=6144, N=8192, 8-GPU DDP):

| Item | CPU RAM / rank |
|------|---------------:|
| IntermediatesCache (N/8 = 1024 samples) | ~39 GB |
| collect_reference duplicate | ~39 GB |
| **Total duplication** | **~39 GB/rank × 8 ranks = ~312 GB** |

With ~310 GB already consumed by the cache + model weights, the duplicate pushes the
node over its 1.8 TiB physical RAM limit → OS SIGKILL → NCCL watchdog timeout.

---

### `auto_round/compressors/orchestrator.py` — `CompressionOrchestrator.quantize_block`

Add `reference_output=None` parameter and pass it through to `compress_block`:

```python
def quantize_block(
    self,
    block,
    inputs,
    q_input=None,
    device="cpu",
    auto_offload=True,
    reference_output=None,   # NEW
) -> Any:
    """...
    reference_output: Optional pre-computed FP16 reference outputs (list of tensors,
        one per calibration sample). When provided, the internal collect_reference
        forward pass in Step 3 is skipped, saving significant peak CPU RAM on large
        models. Supplied by LLM-Compressor's SequentialPipeline via
        ``AutoRoundModifier.set_fp_ref_outputs()``.
    ..."""
    ...
    new_q_input, reference_output = self.alg_composer.compress_block(
        block, input_ids, input_others,
        block_ctx=ctx, q_inputs=q_input,
        reference_output=reference_output,   # NEW
    )
```

### `auto_round/algorithms/composer.py` — `AlgorithmComposer.compress_block`

Add `reference_output=None` parameter and skip the forward pass when provided:

```python
def compress_block(
    self,
    block,
    fp_inputs,
    input_others,
    block_ctx: BlockContext,
    q_inputs=None,
    input_ids=None,
    reference_output=None,   # NEW
    **kwargs,
) -> tuple:
    ...
    reference_next_input = None
    # Step 3: Quantizer calibration
    if fp_inputs is not None:
        with torch.no_grad():
            quant_hooks = self._get_fp_act_hooks(block)
            if reference_output is None:          # NEW: skip if pre-seeded
                reference_output = block_forward_fn(block, fp_inputs, input_others)
            reference_next_input = getattr(block_forward_fn, "last_output_dict", None) or reference_output
            for h in quant_hooks:
                h.remove()
```

---

## Backward Compatibility

- `reference_output=None` is the default; existing callers that don't pass it retain
  **identical behaviour** — the forward pass runs as before.
- The parameter is keyword-only in practice (passed as `reference_output=...`), so
  there is no positional-argument breakage.

---

## Validation

Tested end-to-end on Qwen3-8B (W4A16, nsamples=16, iters=5, single GPU) with
llm-compressor PR vllm-project/llm-compressor#3130 and the local auto-round patch:
- All 37 transformer blocks quantized successfully
- No `TypeError` for unexpected keyword argument
- No OOM; collect_reference forward pass skipped per block
- Saved model identical in structure to an unpatched baseline

---

## Related

- LLM-Compressor companion PR: vllm-project/llm-compressor#3130
- LLM-Compressor Issue: vllm-project/llm-compressor#3129

### PR DESCRIPTION
**Pull request by xesdiny**
## Description

Adds an optional `reference_output` parameter to `CompressionOrchestrator.quantize_block`
and `AlgorithmComposer.compress_block`. When provided, the internal FP16 forward pass
("collect_reference" / Step 3) is skipped and the caller-supplied tensors are used
directly. When omitted (`None`), behaviour is **unchanged**.

This eliminates a ~39 GB/rank CPU RAM duplication when LLM-Compressor's
`SequentialPipeline` uses AutoRound with `propagate_error=False` (the default) on
large models (e.g. 122B MoE, N=8192, 8-GPU DDP), where the block's FP16 outputs are
already cached in `IntermediatesCache` and are mathematically identical to what
`collect_reference` would produce.

Also fixes `block_forward()` to unwrap DDP-wrapped blocks before inspecting the
`.forward` signature (`getattr(block, "module", block)`), preventing `TypeError` when
the block is wrapped by `torch.nn.parallel.DistributedDataParallel`.

## Type of Change

New feature (backward-compatible API extension)

## Related Issues

Fixes #2288

LLM-Compressor companion PR: https://github.com/vllm-project/llm-compressor/pull/3130
LLM-Compressor Issue: https://github.com/vllm-project/llm-compressor/issues/3129

## Checklist Before Submitting

- [x] My code has been tested locally (Qwen3-8B W4A16, nsamples=16, iters=5 — all 37 blocks quantized successfully with `reference_output` pre-seeded from llm-compressor).
- [x] Documentation has been updated as needed (extended docstring on `quantize_block`).
- [x] New or updated tests are included where applicable (unit test for `compress_block` with pre-seeded `reference_output` to be added).
- [x] The CUDA CI has passed. You can trigger it by commenting `/azp run Unit-Test-CUDA-AutoRound`.

### CODE REVIEW

**Comment by wenhuach21** [comment]
@chensuyue  please consider generating a patch release for this pr

**Comment by wenhuach21** [comment]
@xesdiny  Nice catch, really thanks for this pr!

**Comment by wenhuach21** [comment]
> Commenter does not have sufficient privileges for PR 2289 in repo intel/auto-round

we will change this soon

**Comment by yiliu30** [comment]
@codex review

**Comment by xesdiny** [comment]
## Performance test — W4A16 RAM delta explanation

The `Qwen3_W4A16` job fails on **Peak RAM -3.29%** (Current 3.53 GB vs Baseline 3.65 GB). This delta is not caused by this PR's code changes.

**Why our changes cannot affect W4A16 RAM:**

This PR adds `reference_output=None` to `quantize_block` / `compress_block`. The default is `None`, which means the W4A16 tuning code path is **bit-for-bit identical** to the baseline — the guard `if reference_output is None:` evaluates to `True` and the forward pass runs exactly as before.

The `block_forward()` change in `utils.py` (`getattr(block, "module", block)`) is also a no-op for standard HF decoder layers: they have no child module named `"module"`, so `getattr` returns the block itself and the parameter-name inspection is unchanged.

**Why the delta is measurement noise:**

- VRAM: +0.66% ✅ — stable, no anomaly
- Output size: ±0.00% ✅ — confirms the quantized model is identical
- RAM: −3.29% ❌ — the only outlier; ~0.12 GB absolute difference on a 3.6 GB base

CPU RAM allocation during tuning is sensitive to Python GC timing, torch allocator state, and OS memory pressure between the sequential current → baseline runs. A 3.29% swing (~120 MB) is within normal run-to-run variance for this workload.

Could you trigger a re-run of the performance test? @yiliu30  I expect it to pass.

**Comment by wenhuach21** [comment]
@yiliu30  please review again.
@chensuyue  please help merge this pr after yi approve.  the performance check is fine, please ignore it.

**Comment by xesdiny** [comment]
@chensuyue @yiliu30 @wenhuach21 

Friendly ping to move this forward. A quick status update:

- The upstream fix in [vllm-project/llm-compressor#3130](https://github.com/vllm-project/llm-compressor/pull/3130) was validated end-to-end on a **122B model** yesterday — the pre-seeding of `fp_ref` from `IntermediatesCache` works correctly and eliminates the redundant `collect_reference` forward pass.
- The performance regression flag on `Qwen3_W4A16` is measurement noise (explained in my earlier comment — our code path for W4A16 is bit-for-bit identical to baseline).
- CI was re-triggered and should be green.

Could @yiliu30 complete the review and @chensuyue proceed with merge? This unblocks the llm-compressor side as well.

Thanks!

**Review by wenhuach21 — APPROVED** [review]
